src/sage/schemes/hyperelliptic_curves/hyperelliptic_rational_field.py

```diff
@@ -20,21 +20,48 @@ class HyperellipticCurve_rational_field(hyperelliptic_generic.HyperellipticCurve
                                         ProjectivePlaneCurve_field):
 
     def matrix_of_frobenius(self, p, prec=20):
+        """
+        Compute the matrix of Frobenius on Monsky-Washnitzer cohomology using
+        the `p`-adic field with precision ``prec``.
+
+        This function is essentially a wrapper function of
+        :meth:`sage.schemes.hyperelliptic_curves.monsky_washnitzer.matrix_of_frobenius_hyperelliptic`.
+
+        INPUT:
 
-        # BUG: should get this method from HyperellipticCurve_generic
-        def my_chage_ring(self, R):
-            from .constructor import HyperellipticCurve
-            f, h = self._hyperelliptic_polynomials
-            y = self._printing_ring.gen()
-            x = self._printing_ring.base_ring().gen()
-            return HyperellipticCurve(f.change_ring(R), h, "%s,%s" % (x,y))
+        - ``p`` (prime integer or pAdic ring / field ) -- if ``p`` is an integer,
+          constructs a ``pAdicField`` with ``p`` to compute the matrix of
+          Frobenius, otherwise uses the supplied pAdic ring or field.
 
+        - ``prec`` (optional) -- if ``p`` is an prime integer, the `p`-adic
+          precision of the coefficient ring constructed
+
+        EXAMPLES::
+
+            sage: K = pAdicField(5, prec=3)
+            sage: R.<x> = QQ['x']
+            sage: H = HyperellipticCurve(x^5 - 2*x + 3)
+            sage: H.matrix_of_frobenius(K)
+            [            4*5 + O(5^3)       5 + 2*5^2 + O(5^3) 2 + 3*5 + 2*5^2 + O(5^3)     2 + 5 + 5^2 + O(5^3)]
+            [      3*5 + 5^2 + O(5^3)             3*5 + O(5^3)             4*5 + O(5^3)         2 + 5^2 + O(5^3)]
+            [    4*5 + 4*5^2 + O(5^3)     3*5 + 2*5^2 + O(5^3)       5 + 3*5^2 + O(5^3)     2*5 + 2*5^2 + O(5^3)]
+            [            5^2 + O(5^3)       5 + 4*5^2 + O(5^3)     4*5 + 3*5^2 + O(5^3)             2*5 + O(5^3)]
+
+        You can also pass directly a prime `p` with to construct a pAdic field with precision
+        ``prec``::
+
+            sage: H.matrix_of_frobenius(3, prec=2)
+            [        O(3^2)     3 + O(3^2)         O(3^2)         O(3^2)]
+            [    3 + O(3^2)         O(3^2)         O(3^2) 2 + 3 + O(3^2)]
+            [  2*3 + O(3^2)         O(3^2)         O(3^2)    3^-1 + O(3)]
+            [        O(3^2)         O(3^2)     3 + O(3^2)         O(3^2)]
+        """
         import sage.schemes.hyperelliptic_curves.monsky_washnitzer as monsky_washnitzer
         if isinstance(p, (sage.rings.abc.pAdicField, sage.rings.abc.pAdicRing)):
             K = p
         else:
```

Comment:
We probably want to check that `p` is a prime integer

Reply:
I was just relying on the underlying construction to return an error in this case rather than handling it in this call. I can add the extra check though?

Reply:
Either way is fine, my personal preference is the check to be done when needed

Reply:
I think I'll leave it and allow the field constructor to catch the bad p or prec inputs as that check is done anyway